libcxx/include/__algorithm/stable_sort.h

```diff
@@ -189,13 +198,35 @@ struct __stable_sort_switch {
   static const unsigned value = 128 * is_trivially_copy_assignable<_Tp>::value;
 };
 
-template <class _AlgPolicy, class _Compare, class _RandomAccessIterator>
-void __stable_sort(_RandomAccessIterator __first,
-                   _RandomAccessIterator __last,
-                   _Compare __comp,
-                   typename iterator_traits<_RandomAccessIterator>::difference_type __len,
-                   typename iterator_traits<_RandomAccessIterator>::value_type* __buff,
-                   ptrdiff_t __buff_size) {
+template <class _Tp, class = void>
+struct __radix_sort_min_switch {
+  static const unsigned value = (1 << 10);
+};
+
+template <class _Int8>
+struct __radix_sort_min_switch<_Int8, __enable_if_t<is_integral<_Int8>::value && sizeof(_Int8) == 1> > {
+  static const unsigned value = (1 << 8);
+};
+
+template <class _Tp, class = void>
+struct __radix_sort_max_switch {
+  static const unsigned value = (1 << 16);
+};
+
+template <class _Int64>
+struct __radix_sort_max_switch<_Int64, __enable_if_t<is_integral<_Int64>::value && sizeof(_Int64) == 8> > {
+  static const unsigned value = (1 << 15);
+};
```

Comment:
Why are these values not simply in the function and how did you determine these values?

Reply:
It was an analogy to `__stable_sort_switch`.
Values were determined empirically based on benchmarks of raw comparison of radix sort vs `std::sort` vs `std::stable_sort`, and with respect to special benchmark cases such as ascending, descending, pipe-organ and others.

The raw comparison is here
- https://github.com/izvolov/burst?tab=readme-ov-file#%D1%86%D0%B5%D0%BB%D0%BE%D1%87%D0%B8%D1%81%D0%BB%D0%B5%D0%BD%D0%BD%D1%8B%D0%B5-%D1%81%D0%BE%D1%80%D1%82%D0%B8%D1%80%D0%BE%D0%B2%D0%BA%D0%B8
- https://github.com/izvolov/burst/blob/master/doc/README.md#intsort
Code:
- https://github.com/izvolov/burst/blob/master/benchmark/burst/algorithm/integer_sort_comparison.py.in
- https://github.com/izvolov/burst/blob/master/benchmark/burst/algorithm/radix_sort.cpp

Reply:
I think I'd make them `constexpr` functions instead, since we can.

Reply:
Do you mean something like this?
```cpp
template <class _Tp>
constexpr unsigned __radix_sort_min_bound () {
  static_assert(std::is_integral_v<_Tp>);
  if constexpr (sizeof(_Tp) == 1) {
      return 1 << 8;
  }

  return 1 << 10;
}
```


Is is legal to use C++11 and later features in `stable_sort.h` since there aren't any version guards there?
If it is, I'd also rewrite the code to remove `_BoolConstant<_EnableRadixSort>`.

Reply:
Done.
The only my concern is that now we require C++17 to enable radix sort. Is it ok?

Reply:
`if constexpr` is C++17 AFAIK.

Reply:
Yes, but we don't care about pedantic warnings. All compilers we care about support it in C++11 and 14 mode: https://godbolt.org/z/YGrsh9Kh6

Reply:
Hmm, I tried it at first, but got CI errors. So I made a commit with C++17 guard: https://github.com/llvm/llvm-project/pull/104683/commits/eb84acf2e89f7b7454147b1af3851683eacf4e64